PmwOut: Add method to enable/disable PWM

hugueskamba opened the pull request:
### Description
It is now possible to temporarily suspend PWM and safely preserve the duty
cycle set. This functionality is needed to allow a device to enter deep
sleep as a PWM instance prevents deep sleep in order for the timer it
relies on to run so its output can be modified. The duty cycle configuration
can be restored upon resuming from deep sleep.
<!--
    Required
    Add here detailed changes summary, testing results, dependencies
    Good example: https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html (Pull request template)
-->


### Pull request type

<!--
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [X] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers
@evedon @bulislaw @kjbracey-arm 
<!--
    Optional
    Request additional reviewers with @username
-->

### Release Notes

<!--
    Optional
    In case of breaking changes, functionality changes or refactors, please add release notes here. 
    For more information, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html#pull-request-types).
-->

bulislaw commented on drivers/source/PwmOut.cpp:
```diff
@@ -98,6 +98,19 @@ void PwmOut::pulsewidth_us(int us)
     core_util_critical_section_exit();
 }
 
+void PwmOut::enable_pwm(bool enabled)
```
Do we actually need to `pwmout_free` to free the HW (and then to restore the state)? I'm not sure what will happen with HW state if the high frequency clock would go down and also HW may limit the low power state based on active HW.

hugueskamba replied:
https://github.com/ARMmbed/mbed-os/pull/11641#issuecomment-539477808

hugueskamba committed:
PmwOut: Add methods to suspend and resume PWM

It is now possible to temporarily suspend PWM and safely preserve the duty
cycle set. This functionality is needed to allow a device to enter deep
sleep as a PWM instance prevents deep sleep in order for the timer it
relies on to run so its output can be modified. The duty cycle configuration
can be restored upon resuming from deep sleep.

bulislaw reviewed (approved):
Looks good. Not related to this PR but this class forces users to use floats 😞

evedon commented on drivers/source/PwmOut.cpp:
```diff
@@ -98,6 +98,26 @@ void PwmOut::pulsewidth_us(int us)
+
+void PwmOut::resume()
+{
+    core_util_critical_section_exit();
```
Should be core_util_critical_section_enter()

hugueskamba replied:
Corrected [here](https://github.com/ARMmbed/mbed-os/compare/1645cc774b8a27fad92693314e16001133578e25..fdc76051f30588bac2df8155c26ba6cd700b5879).
Thanks.

evedon reviewed (dismissed):
We also need to add greentea tests  (is there an existing test suite that can be expanded)?

hugueskamba committed:
PwmOut: Add unit test

evedon commented on UNITTESTS/drivers/PwmOut/test_pwmout.cpp:
```diff
@@ -0,0 +1,202 @@
+    MOCK_METHOD0(sleep_manager_unlock_deep_sleep_internal, void());
+    MOCK_METHOD0(sleep_manager_lock_deep_sleep_internal, void());
+
+    static MockMbedPowerMgmt *get_instance(bool create_instance)
```
Why do you need to keep a ref_counter?
Create a new instance only when `instance == nullptr`

hugueskamba replied:
Because it is good coding practice to have one when implementing a singleton class with a method to release the instance. As one does not want to release an instance in a singleton if it is still referenced.

evedon replied:
This function should return a pointer to the static instance i.e. it should not be possible to return `NULL`. If `create_instance` is false, the function can return NULL. Also if you are keeping a reference count I don't see why this is not matching in `release_instance`.

hugueskamba replied:
Because of what I mentioned [here](https://github.com/ARMmbed/mbed-os/pull/11641#discussion_r340864746), I specifically want it to return a null pointer if the call came from instantiating the `PwmOut` object in `Setup` so the mock object expectation count does not increase.

I don't understand what you mean by `..why this not matching in release_instance`.  `release_instance` decrements the counter and checks that it is zero before delete the instance.

evedon replied:
In a singleton pattern, it is expected that a call to get_instance() does not return NULL as it returns a pointer to a static instance. So this is counterintuitive.

In addition, this extra complexity (passing a `create_instance` argument and checking references) does not provide value to the code under test; at best it would ensure that the test framework is not leaking memory but it is not able to do so because  the instance is deleted in the teardown call.

> Also if you are keeping a reference count I don't see why this is not matching in release_instance.
Sorry this is extremely badly worded. I meant to say that some calls to `get_instance` increment `ref_counter` but all calls to `release_instance` decrement the counter. Again, I fell this is counterintuitive.

hugueskamba replied:
The following force-pushes address what is mentioned in your last comment.
*  https://github.com/ARMmbed/mbed-os/compare/db1c269192cd81824596368ae197007ff4266d4b..ab2544b04d2ff54af0c9ae8d26bad2f5028c7132
*  https://github.com/ARMmbed/mbed-os/compare/ab2544b04d2ff54af0c9ae8d26bad2f5028c7132..8bcddc487357de00471d299ea00842fa12506638

evedon commented on UNITTESTS/drivers/PwmOut/test_pwmout.cpp:
```diff
@@ -0,0 +1,202 @@
+    // by the teardown
+    Mock::VerifyAndClearExpectations(mbed_mgmt_mock);
+    // There will not be a leak since the destructor releases the singleton
+    Mock::AllowLeak(mbed_mgmt_mock);
```
If there is no leak, `Mock::AllowLeak` does not seem correct?

hugueskamba replied:
This method was specifically implemented for cases where the testing framework cannot  detect that a mocked object has been safely destroyed. I destroy the singleton instance in the teardown so it is fine.

The leak error which is suppressed by `AllowLeak` occurs because I added `VerifyAndClearExpectations` so the expectation assertion does not include the call in the `PwmOut` destructor found in the teardown.

evedon replied:
I was confused that the comment says "There will not be a leak" and the code  "Mock::AllowLeak".
It seemed that one of them was wrong.

Perhaps you could slighlty rework  "Suppress the memory leak error. The mock singleton will be released and deleted in the TearDown call."

hugueskamba replied:
[Done](https://github.com/ARMmbed/mbed-os/compare/8bcddc487357de00471d299ea00842fa12506638..6352a011849a79d8ad6d55b91d7c5d9a2e8ae1d1).

evedon commented on UNITTESTS/drivers/PwmOut/test_pwmout.cpp:
```diff
@@ -0,0 +1,202 @@
+        // Create a mock object singleton after the PwmOut object
+        // instantiation so the sleep_manager_lock_deep_sleep call by the
+        // constructor doesnot increment the call count.
+        // Now it is ok because a test case is about to start.
```
But we should also test that at creation the PwmOut object calls sleep_manager_lock_deep_sleep.
Can you check at the beginning of the test case that the deep_sleep lock is held and then run your test case?

hugueskamba replied:
>@jamesbeyond suggested that we manually run the FPGA test shield tests which test pwm.
In addition we should add tests for the new suspend/resume API in https://github.com/ARMmbed/mbed-os/blob/master/TESTS/mbed_hal_fpga_ci_test_shield/pwm/main.cpp

I added tests for `resume` and `suspend` not for the constructor.

evedon replied:
But you are adding a lot of complexity in order not to test something that adds value...

hugueskamba replied:
If we want to test the constructor a separate test case should be added. 

The added test cases are for the two methods I mentioned. The expectations in the test cases are clear without having to explain anything with a comment (`// Add an extra expected call because the first call is by the constructor`). 
What if the constructor changes and no longer calls the deep sleep function in the future? We will have to come and change the expectation number again for the `resume` and `suspend` test cases even if those two methods have not been changed. We will have a dependency on the constructor. This way it is fixed as they only check the call counts by `resume`/`suspend.`

evedon commented on UNITTESTS/drivers/PwmOut/test_pwmout.cpp:
```diff
@@ -0,0 +1,202 @@
+    // case.
+    MockMbedPowerMgmt *mbed_mgmt_mock = MockMbedPowerMgmt::get_instance(false);
+    if (mbed_mgmt_mock != nullptr) {
+        // Call only if the singleton  was created by a test case.
```
There is a lot of added complexity which does not seem necessary. Why would this function be called outside of a test case? 
`MockMbedPowerMgmt::get_instance` should take no argument and always return a pointer to the static instance so you could do `MockMbedPowerMgmt::get_instance()->sleep_manager_lock_deep_sleep_internal();`

hugueskamba replied:
It will be far more constructive to ask why something was done a certain way in order to understand the need and potentially suggest a simpler way (if one exists) instead of saying that it is unnecessary or how something should or should not be.

Why would the function be called outside of a test case? Because the `PwmOut` object is instantiated in `Setup` and the constructor calls `sleep_manager_lock_deep_sleep`.

Why is it created in `Setup`? So the expectation count in the test cases does not include the call by the constructor. The test cases are for unit tests, to check the call counts by the `resume` and `suspend` units only. 

Why does `get_instance` gets a parameter? So the call by the constructor/destructor does not instantiate the singleton instance (and start incrementing the expectation call count) if one was not already created for the purpose of a test case and start incrementing the count. 

In other words without what you mention `test_suspend` would look as follows:

```C
TEST_F(TestPwmOut, test_suspend)
{
    using ::testing::Mock;
    // I EXPECT IT TO BE CALLED ONLY ONCE NOT TWICE, IT LOOKS WRONG
    // I DONT REALLY CARE WHAT THE CONSTRUCTOR IS DOING HERE.
    EXPECT_CALL(
        *mbed_mgmt_mock, sleep_manager_unlock_deep_sleep_internal()
    ).Times(2); 
    pwm_obj->suspend();
    pwm_obj->suspend();

    // Force gMock to verify a mock object singleton before it is destructed
    // by the teardown
    Mock::VerifyAndClearExpectations(mbed_mgmt_mock);
    // There will not be a leak since the destructor releases the singleton
    Mock::AllowLeak(mbed_mgmt_mock);
}
```
Instead of what it is now:

```C
TEST_F(TestPwmOut, test_suspend)
{
    using ::testing::Mock;

    EXPECT_CALL(
        *mbed_mgmt_mock, sleep_manager_unlock_deep_sleep_internal()
    ).Times(1);
    pwm_obj->suspend();
    pwm_obj->suspend();

    // Force gMock to verify a mock object singleton before it is destructed
    // by the teardown
    Mock::VerifyAndClearExpectations(mbed_mgmt_mock);
    // There will not be a leak since the destructor releases the singleton
    Mock::AllowLeak(mbed_mgmt_mock);
}
```

If we are interested in testing the constructor then I would have another test like this:

```C
TEST_F(TestPwmOut, test_constructor)
{
    using ::testing::Mock;

    EXPECT_CALL(
        *mbed_mgmt_mock, sleep_manager_lock_deep_sleep_internal()
    ).Times(1);
    PwmOut pwm(PTC1);

    // Force gMock to verify a mock object singleton before it is destructed
    // by the teardown
    Mock::VerifyAndClearExpectations(mbed_mgmt_mock);
    // There will not be a leak since the destructor releases the singleton
    Mock::AllowLeak(mbed_mgmt_mock);
}
```

evedon commented on UNITTESTS/drivers/PwmOut/test_pwmout.cpp:
```diff
@@ -0,0 +1,202 @@
+    static MockMbedPowerMgmt *get_instance(bool create_instance)
+    {
+        if (instance == nullptr && create_instance) {
+            instance = new MockMbedPowerMgmt();
```
You can assert that instance is not NULL.

hugueskamba replied:
Why?

evedon replied:
It's a suggestion to ensure that you are not running out of memory. But as this is not widely used, let's ignore.

evedon commented on UNITTESTS/drivers/PwmOut/test_pwmout.cpp:
```diff
@@ -0,0 +1,202 @@
+    MockMbedPowerMgmt() {}
+
+    static MockMbedPowerMgmt *instance;
+    static int ref_counter;
```
Add an underscore to private members to follow convention.

hugueskamba replied:
[Done](https://github.com/ARMmbed/mbed-os/compare/f6fb26510d8ca6ff3e617a531b394a41264e66e3..ac77d3a288c6d9816bcb1601f0ef7db30e30813c).

evedon reviewed (approved):
Thanks for the changes. 
It might have been easier to use the framework to clear the expectations before the test starts but this works as well.
